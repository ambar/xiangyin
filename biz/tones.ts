/**
 * https://zh.wikipedia.org/wiki/四聲
 */

export const toneCodeByName = {
  T0: '轻声',
  T1: '平声',
  T1a: '阴平',
  T1b: '阳平',
  T2: '上声',
  T2a: '阴上',
  T2b: '阳上',
  T3: '去声',
  T3a: '阴去',
  T3b: '阳去',
  T4: '入声',
  T4a: '阴入',
  T4b: '阳入',
}

export const toneNameByCode = {
  轻声: 'T0',
  平: 'T2',
  平声: 'T2',
  阴平: 'T1a',
  阳平: 'T1b',
  上: 'T2',
  上声: 'T2',
  阴上: 'T2a',
  阳上: 'T2b',
  去: 'T3',
  去声: 'T3',
  阴去: 'T3a',
  阳去: 'T3b',
  入: 'T4',
  入声: 'T4',
  阴入: 'T4a',
  阳入: 'T4b',
}

export const csToneNoByCode = {
  T0: '0',
  T1a: '1',
  T1b: '2',
  T2: '3',
  T3a: '4',
  T3b: '5',
  T4: '6',
}

export const csToneNoByTone = {
  '1': 1,
  '2': 2,
  '3': 3,
  '5': 4,
  '6': 5,
  '7': 6,
}
