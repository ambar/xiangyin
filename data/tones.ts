/**
 * 声调格式转换
 *
 * @see https://zh.wikipedia.org/wiki/四聲
 */

export type ToneType = 'ToneName' | 'CSToneNo' | 'OctetToneNo'

export type ToneCode = keyof typeof toneCode2toneName
export const toneCode2toneName = {
  T0: '轻声',
  T1: '平声',
  T1a: '阴平',
  T1b: '阳平',
  T2: '上声',
  T2a: '阴上',
  T2b: '阳上',
  T3: '去声',
  T3a: '阴去',
  T3b: '阳去',
  T4: '入声',
  T4a: '阴入',
  T4b: '阳入',
}

export type ToneName = keyof typeof toneName2ToneCode
export const toneName2ToneCode = {
  轻声: 'T0',
  // 平: 'T2',
  平声: 'T2',
  阴平: 'T1a',
  阳平: 'T1b',
  // 上: 'T2',
  上声: 'T2',
  阴上: 'T2a',
  阳上: 'T2b',
  // 去: 'T3',
  去声: 'T3',
  阴去: 'T3a',
  阳去: 'T3b',
  // 入: 'T4',
  入声: 'T4',
  阴入: 'T4a',
  阳入: 'T4b',
}

/** 长沙话序号声调 */
export type CSToneNo = 1 | 2 | 3 | 4 | 5 | 6

/** 长沙话序号声调 */
export const toneCode2csToneNo: Partial<Record<ToneCode, CSToneNo | ''>> = {
  T0: '',
  T1a: 1,
  T1b: 2,
  T2: 3,
  T3a: 4,
  T3b: 5,
  T4: 6,
  T4a: 6,
}

/** 八位数字声调 */
export type OctetToneNo = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8
/** 长沙话在用的八位数字声调  */
export type CSOctetToneNo = Exclude<OctetToneNo, 4 | 8>

/** 八位数字声调 */
export const octetToneNo2csToneNo: Record<CSOctetToneNo, number> = {
  1: 1,
  2: 2,
  3: 3,
  5: 4,
  6: 5,
  7: 6,
}

export const csOctetToneNo2toneName: Record<CSOctetToneNo, ToneName> = {
  1: '阴平',
  2: '阳平',
  3: '上声',
  5: '阴去',
  6: '阳去',
  7: '入声',
}

/** 四声转长沙话序号 */
export const csToneName2csToneNo: Partial<Record<ToneName, CSToneNo | ''>> = {
  轻声: '',
  阴平: 1,
  阳平: 2,
  上声: 3,
  阴去: 4,
  阳去: 5,
  入声: 6,
}

/** 调值转长沙话序号 */
export const toneValue2csToneNo: Record<number, CSToneNo> = {
  33: 1,
  13: 2,
  41: 3,
  55: 4,
  11: 5,
  24: 6,
}
