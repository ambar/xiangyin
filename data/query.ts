import yitizi from 'yitizi'
import {ToneType} from './tones'
import * as 汉语方音字汇 from './汉语方音字汇'
import * as 湘音检字 from './湘音检字'
import * as 长沙话音档 from './长沙话音档'

export enum Source {
  湘音检字 = '湘音检字',
  汉语方音字汇 = '汉语方音字汇',
  长沙话音档 = '长沙话音档',
}

export const sourceOptions = [
  {value: Source.湘音检字, label: '《湘音检字》1937'},
  {value: Source.长沙话音档, label: '《长沙话音档》1997'},
  {value: Source.汉语方音字汇, label: '《汉语方音字汇》2003'},
]

let queryMap = new Map([
  [Source.湘音检字, 湘音检字.query],
  [Source.长沙话音档, 长沙话音档.query],
  [Source.汉语方音字汇, 汉语方音字汇.queryCS],
])

export const queryPinyin = (
  char: string,
  shouldQueryVariants = false,
  source: Source,
  toneType?: ToneType
) => {
  let query = queryMap.get(source)!
  let result = query(char, toneType)
  if (!result.length && shouldQueryVariants) {
    for (const v of yitizi.get(char)) {
      let r = query(v, toneType)
      if (r.length) {
        result = r
        break
      }
    }
  }
  return result
}

export const queryPinyinAll = (
  char: string,
  shouldQueryVariants = false,
  toneType?: ToneType
): [Source, QueryResult][] => {
  return [...queryMap.keys()].map((k) => [
    k,
    queryPinyin(char, shouldQueryVariants, k, toneType),
  ])
}

export type QueryResult = ReturnType<typeof queryPinyin>
