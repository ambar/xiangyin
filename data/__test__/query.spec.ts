import {queryPinyin, queryPinyinAll, Source} from '../query'
import {ToneType, ToneTypes} from '../tones'

// 閱阅祿禄禿秃續续
test('shouldQueryVariants', () => {
  const groupA = ['閱阅', '祿禄', '禿秃']
  groupA.forEach((chars) => {
    const [a, b] = [...chars]
    expect(queryPinyin(b, false, Source.长沙话音档)).toEqual([])
    expect(queryPinyin(a, true, Source.长沙话音档)).toEqual(
      queryPinyin(b, true, Source.长沙话音档)
    )
  })

  const groupB = ['閱阅', '祿禄', '禿秃', '續续']
  groupB.forEach((chars) => {
    const [a, b] = [...chars]
    expect(queryPinyin(b, false, Source.汉语方音字汇)).toEqual([])
    expect(queryPinyin(a, true, Source.汉语方音字汇)).toEqual(
      queryPinyin(b, true, Source.汉语方音字汇)
    )
  })
})

test.each(ToneTypes.concat('XYZ' as ToneType))(`toneType: %s`, (toneType) => {
  expect(queryPinyinAll('入', false, {toneType})).toMatchSnapshot()
})

test('XPA', () => {
  Array.from('風吉辰人卻').map((c) => {
    expect(
      queryPinyinAll(c, false, {pinyinType: 'XPA'}).map(([s, r]) => [
        s,
        c,
        r.map((x) => x.音),
      ])
    ).toMatchSnapshot()
  })
})
