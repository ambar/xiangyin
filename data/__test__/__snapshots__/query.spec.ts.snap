// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XPA 1`] = `
Array [
  Array [
    "湘音检字",
    "風",
    Array [
      "hong",
      "hong",
    ],
  ],
  Array [
    "长沙话音档",
    "風",
    Array [
      "hong",
    ],
  ],
  Array [
    "汉语方音字汇",
    "風",
    Array [
      "hen",
    ],
  ],
]
`;

exports[`XPA 2`] = `
Array [
  Array [
    "湘音检字",
    "吉",
    Array [
      "ji",
    ],
  ],
  Array [
    "长沙话音档",
    "吉",
    Array [
      "ji",
    ],
  ],
  Array [
    "汉语方音字汇",
    "吉",
    Array [
      "ji",
    ],
  ],
]
`;

exports[`XPA 3`] = `
Array [
  Array [
    "湘音检字",
    "辰",
    Array [
      "shen",
    ],
  ],
  Array [
    "长沙话音档",
    "辰",
    Array [
      "shen",
    ],
  ],
  Array [
    "汉语方音字汇",
    "辰",
    Array [
      "sen",
    ],
  ],
]
`;

exports[`XPA 4`] = `
Array [
  Array [
    "湘音检字",
    "人",
    Array [
      "ren",
    ],
  ],
  Array [
    "长沙话音档",
    "人",
    Array [
      "ren",
    ],
  ],
  Array [
    "汉语方音字汇",
    "人",
    Array [
      "ren",
    ],
  ],
]
`;

exports[`XPA 5`] = `
Array [
  Array [
    "湘音检字",
    "卻",
    Array [
      "qio",
    ],
  ],
  Array [
    "长沙话音档",
    "卻",
    Array [
      "qio",
    ],
  ],
  Array [
    "汉语方音字汇",
    "卻",
    Array [],
  ],
]
`;

exports[`toneType: CSToneNo 1`] = `
Array [
  Array [
    "湘音检字",
    Array [
      Object {
        "声": "",
        "调": 6,
        "释": "出～",
        "音": "y",
        "韵": "y",
      },
      Object {
        "声": "ȵ",
        "调": 4,
        "释": "",
        "音": "ȵiɑ",
        "韵": "iɑ",
      },
    ],
  ],
  Array [
    "长沙话音档",
    Array [
      Object {
        "声": "",
        "调": 6,
        "释": "",
        "音": "y",
        "韵": "y",
      },
    ],
  ],
  Array [
    "汉语方音字汇",
    Array [
      Object {
        "声": "",
        "调": 6,
        "释": "",
        "音": "y",
        "韵": "y",
      },
    ],
  ],
]
`;

exports[`toneType: OctetToneNo 1`] = `
Array [
  Array [
    "湘音检字",
    Array [
      Object {
        "声": "",
        "调": 7,
        "释": "出～",
        "音": "y",
        "韵": "y",
      },
      Object {
        "声": "ȵ",
        "调": 5,
        "释": "",
        "音": "ȵiɑ",
        "韵": "iɑ",
      },
    ],
  ],
  Array [
    "长沙话音档",
    Array [
      Object {
        "声": "",
        "调": 7,
        "释": "",
        "音": "y",
        "韵": "y",
      },
    ],
  ],
  Array [
    "汉语方音字汇",
    Array [
      Object {
        "声": "",
        "调": 7,
        "释": "",
        "音": "y",
        "韵": "y",
      },
    ],
  ],
]
`;

exports[`toneType: ToneLetter 1`] = `
Array [
  Array [
    "湘音检字",
    Array [
      Object {
        "声": "",
        "调": "˨˦",
        "释": "出～",
        "音": "y",
        "韵": "y",
      },
      Object {
        "声": "ȵ",
        "调": "˥˥",
        "释": "",
        "音": "ȵiɑ",
        "韵": "iɑ",
      },
    ],
  ],
  Array [
    "长沙话音档",
    Array [
      Object {
        "声": "",
        "调": "˨˦",
        "释": "",
        "音": "y",
        "韵": "y",
      },
    ],
  ],
  Array [
    "汉语方音字汇",
    Array [
      Object {
        "声": "",
        "调": "˨˦",
        "释": "",
        "音": "y",
        "韵": "y",
      },
    ],
  ],
]
`;

exports[`toneType: ToneName 1`] = `
Array [
  Array [
    "湘音检字",
    Array [
      Object {
        "声": "",
        "调": "入声",
        "释": "出～",
        "音": "y",
        "韵": "y",
      },
      Object {
        "声": "ȵ",
        "调": "阴去",
        "释": "",
        "音": "ȵiɑ",
        "韵": "iɑ",
      },
    ],
  ],
  Array [
    "长沙话音档",
    Array [
      Object {
        "声": "",
        "调": "入聲",
        "释": "",
        "音": "y",
        "韵": "y",
      },
    ],
  ],
  Array [
    "汉语方音字汇",
    Array [
      Object {
        "声": "",
        "调": "入声",
        "释": "",
        "音": "y",
        "韵": "y",
      },
    ],
  ],
]
`;

exports[`toneType: ToneValue 1`] = `
Array [
  Array [
    "湘音检字",
    Array [
      Object {
        "声": "",
        "调": 24,
        "释": "出～",
        "音": "y",
        "韵": "y",
      },
      Object {
        "声": "ȵ",
        "调": 55,
        "释": "",
        "音": "ȵiɑ",
        "韵": "iɑ",
      },
    ],
  ],
  Array [
    "长沙话音档",
    Array [
      Object {
        "声": "",
        "调": 24,
        "释": "",
        "音": "y",
        "韵": "y",
      },
    ],
  ],
  Array [
    "汉语方音字汇",
    Array [
      Object {
        "声": "",
        "调": 24,
        "释": "",
        "音": "y",
        "韵": "y",
      },
    ],
  ],
]
`;

exports[`toneType: XYZ 1`] = `
Array [
  Array [
    "湘音检字",
    Array [
      Object {
        "声": "",
        "调": 6,
        "释": "出～",
        "音": "y",
        "韵": "y",
      },
      Object {
        "声": "ȵ",
        "调": 4,
        "释": "",
        "音": "ȵiɑ",
        "韵": "iɑ",
      },
    ],
  ],
  Array [
    "长沙话音档",
    Array [
      Object {
        "声": "",
        "调": 6,
        "释": "",
        "音": "y",
        "韵": "y",
      },
    ],
  ],
  Array [
    "汉语方音字汇",
    Array [
      Object {
        "声": "",
        "调": 6,
        "释": "",
        "音": "y",
        "韵": "y",
      },
    ],
  ],
]
`;
